sdks/python/apache_beam/transforms/external_test.py

```diff
@@ -48,6 +59,152 @@
 # pylint: enable=wrong-import-order, wrong-import-position
 
 
+def get_payload(args):
+  return ExternalConfigurationPayload(configuration=args).SerializeToString()
+
+
+class PayloadBase(object):
+  values = {
+      'integer_example': 1,
+      'string_example': u'thing',
+      'list_of_strings': [u'foo', u'bar'],
+      'optional_kv': (u'key', 1.1),
+      'optional_integer': None,
+  }
+
+  bytes_values = {
+      'integer_example': 1,
+      'string_example': 'thing',
+      'list_of_strings': ['foo', 'bar'],
+      'optional_kv': ('key', 1.1),
+      'optional_integer': None,
+  }
+
+  args = {
+      'integer_example': ConfigValue(
+          coder_urn=['beam:coder:varint:v1'],
+          payload=VarIntCoder().encode(values['integer_example'])),
+      'string_example': ConfigValue(
+          coder_urn=['beam:coder:string_utf8:v1'],
+          payload=StrUtf8Coder().encode(values['string_example'])),
```

Comment:
This is not sufficient because the Python `StrUtf8Coder` does not add a length prefix while the Java one does. Maybe that is actually a bug. The old version was using a length prefix coder before the bytes coder.

Reply:
See https://github.com/apache/beam/blob/9678149872de2799ea1643f834f2bec88d346af8/sdks/python/apache_beam/coders/coders.py#L326

Reply:
Difference between Python and Java UTF-8 coders was discussed before in dev list: https://lists.apache.org/thread.html/a3a0d9e7c4196bb6be14ba0bec103209317dcb98e781560eb3ccd48c@%3Cdev.beam.apache.org%3E

Unfortunately I don't believe we resolved this though.

Reply:
This is the nested vs. unnested issue back to bite us again. 

Here we should be using the nested context consistently. For Python, write `coder.get_impl().encode_nested(value)`. Probably would be worth adding an `encode_nested` method to coder itself.

Reply:
@robertwb  Can you provide a bit more explanation for the beam-newbs in the audience (me), please?  Does this just ensure lengths prefixed for applicable types?

Reply:
Basically, ever coder actually represents two concrete encodings, a "nested" one for use in an unbounded input stream (e.g. as a stream of many elements, or within a composite element type like list), and a "outer" one for use when the end-of-record must be provided explicitly (e.g. when writing to a text file, were the newline delimits strings and prefixing each line with a string length would be bad). (In Java this is reified in the second Context argument of the encode/decode functions. For Python it's only visible in the impl layer). 

For some coders, nested and unnested are identical (e.g. varints or doubles, where the element length is implicit in the encoding). For others, a length prefix is added (e.g. utf8 strings and bytes). For others still, it gets pushed down (e.g. for kv coders, the key is always encoded nested, and the value is encoded in the outer context). 

We're trying to move to use the "nested" one everywhere, as this is really confusing (see the linked thread).

Reply:
I think we should be good here and elsewhere to use the nested encoding. On the Java side, we need to change the `decode` call in ExpansionServer to use the `NESTED` parameter. The old code here did not need that because it simply wrapped everything in an additional LengthPrefixCoder, which `encode_nested` will essentially do as well. So either reverting back to adding a wrapping LengthPrefixCoder or using encode_nested should work. Note that the latter also requires the additional NESTED parameter on the Java side in ExpansionServer.